### PR TITLE
umicollapse module: Set initial heap size and modificaitons to log file creation

### PR DESCRIPTION
**Pull request by MatthiasZepper**
This PR suggests two changes to the upstream nf-core module that I had to introduce to make it work [within an evaluation version the rnaseq pipeline using this module](https://github.com/MatthiasZepper/rnaseq/tree/eval_umicollapse).

The first modification adds a `-Xmx` parameter to the command that specifies the maximum heap size of the virtual machine. The parameter is used exactly as it is found in several other nf-core modules wrapping Java tools.

The second modification writes a logfile from stdout, since moving the `.command.log` doesn't seem to be a valid approach. The file seemingly does not yet exist in the work directory when the command is about to be executed.

For me, it was fine to not set an explicit stack size with `-Xss`, but @grst [apparently had to do it for the smrnaseq pipeline](https://github.com/nf-core/smrnaseq/blob/d5192d48a3491504b41fe1af51860d1d9773c632/modules/nf-core/umicollapse/main.nf), so maybe it would be helpful to add this as well?

The additional label `process_high` was added because module lint complained about the lack of standard process labels. If `process_high_memory` is declared after `process_high` in the config, i[t will take precedence over the memory setting](https://github.com/MatthiasZepper/Nextflow_withLabel_prio). 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by grst** [comment]
> For me, it was fine to not set an explicit stack size with -Xss, but @grst [apparently had to do it for the smrnaseq pipeline](https://github.com/nf-core/smrnaseq/blob/d5192d48a3491504b41fe1af51860d1d9773c632/modules/nf-core/umicollapse/main.nf), so maybe it would be helpful to add this as well?

Yeah, we did get a stackoverflow error without that. Unfortunately, this can't be passed directly to the `umicollapse` command (which turned out to be a python wrapper script generated by conda), which is why I ended up calling the `jar` directly: 
https://github.com/nf-core/smrnaseq/blob/d5192d48a3491504b41fe1af51860d1d9773c632/modules/nf-core/umicollapse/main.nf#L27-L34

**Comment by MatthiasZepper** [comment]
> Yeah, we did get a stackoverflow error without that. Unfortunately, this can't be passed directly to the `umicollapse` command (which turned out to be a python wrapper script generated by conda), which is why I ended up calling the `jar` directly: https://github.com/nf-core/smrnaseq/blob/d5192d48a3491504b41fe1af51860d1d9773c632/modules/nf-core/umicollapse/main.nf#L27-L34

I saw it, but wondered if the path `/usr/local/share/umicollapse-${VERSION}/umicollapse.jar` is compatible with conda? 

Also, I wondered if hardcoding the stack size to 99M advisable? Should that not also depend on the task memory? I am not good at Groovy, but maybe choosing some fraction of it, like `-Xss${String.format("%.0f", Math.rint(0.2 * task.memory.toGiga()))}g` could work?

**Comment by grst** [comment]
> I saw it, but wondered if the path /usr/local/share/umicollapse-${VERSION}/umicollapse.jar is compatible with conda?

good point, I just tried in the container. But it should be possible to get this path somehow, maybe
```
$(dirname $(which umicollapse))/../share/umicollapse-${VERSION}.jar
```

Definitely a good idea to make this dependent on the task memory. We are still debugging tbh.

**Comment by grst** [comment]
We had to increase the stack size even further. By setting it to `1G` all our jobs succeeded. So setting this to a few percent of the heapspace probably makes sense. 

I pushed a commit trying this out and also trying to obtain the jar file independent of wheather we are using conda or a container.

**Comment by grst** [comment]
ok, seems to work in principle. The tests fail because of some checksum mismatch

File: modules/nf-core/umicollapse/main.nf [hunk 1]
```diff
@@ -33,7 +33,7 @@ process UMICOLLAPSE {
     # by conda that allows to set the heap size (Xmx), but not the stack size (Xss).
     # `which` allows us to get the directory that contains `umicollapse`, independent of wheather we
     # are in a container or conda environment.
-    UMICOLLAPSE_JAR=$(dirname $(which umicollapse))/../share/umicollapse-${VERSION}/umicollapse.jar
+    UMICOLLAPSE_JAR=\$(dirname \$(which umicollapse))/../share/umicollapse-${VERSION}/umicollapse.jar
```
Comment (MatthiasZepper) [inline modules/nf-core/umicollapse/main.nf:36]: Don't you have to escape the `$` for `${Version}` as well then?

Also mind that I had already added the `-Xmx` parameter, so you need to remove my declaration then.

File: modules/nf-core/umicollapse/main.nf [hunk 2]
```diff
@@ -21,16 +22,27 @@ process UMICOLLAPSE {
...
+    # by conda that allows to set the heap size (Xmx), but not the stack size (Xss).
+    # `which` allows us to get the directory that contains `umicollapse`, independent of wheather we
+    # are in a container or conda environment.
+    UMICOLLAPSE_JAR=\$(dirname \$(which umicollapse))/../share/umicollapse-${VERSION}/umicollapse.jar
+    java \\
+        -Xmx${max_heap_size_mega}M \\
+        -Xss${max_stack_size_mega}M \\
+        -jar \$UMICOLLAPSE_JAR \\
         bam \\
+        -Xmx${task.memory.toGiga()}g  \\
```
Comment (grst) [inline modules/nf-core/umicollapse/main.nf:42]:
```suggestion
```

File: modules/nf-core/umicollapse/main.nf (hunk 1 above)
Comment (grst) [reply modules/nf-core/umicollapse/main.nf:36]: VERSION is a groovy variable, so no ;) 
But good catch with the double Xmx!

**Comment by apeltzer** [comment]
I know this might become a rabbit hole by itself, but should we look into migrating pytest to `nf-test`?

**Comment by grst** [comment]
We totally should. The tests are already broken, no point in fixing pytest. 
@MatthiasZepper, do you want to go for it? Otherwise I can give it a shot

**Comment by MatthiasZepper** [comment]
> We totally should. The tests are already broken, no point in fixing pytest. @MatthiasZepper, do you want to go for it? Otherwise I can give it a shot

I have 0 experience with nf-test and not really the capacity to familiarize myself with it at the moment. So if you volunteer, I am more than grateful to accept that generous offer.

**Comment by apeltzer** [comment]
I have started working on nf-test tests, we can continue on the branch here on modules as we can all push there directly and I'm lacking permissions on your repo @MatthiasZepper

**Comment by apeltzer** [comment]
Opened this one for collaborative editing: https://github.com/nf-core/modules/pull/4795

**Pull request closed without merging**